gix-worktree/src/diff.rs

```diff
@@ -0,0 +1,292 @@
+            }
+        };
+        self.data_changed = hash_changed;
+        Ok(())
+    }
+}
+
+#[allow(missing_docs)]
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+    #[error("Could not convert path to UTF8 {path}")]
+    IllformedUtf8 { path: BString },
+    #[error("The clock was off when reading file related metadata after updating a file on disk")]
+    Time(#[from] std::time::SystemTimeError),
+    #[error("IO error while writing blob or reading file metadata or changing filetype")]
+    Io(#[from] io::Error),
+}
+
+#[derive(Clone, Debug)]
+/// A change between the index and the worktree computed by [`compate_to_index`]
+pub struct Change<'a> {
+    /// The index entry that changed
+    pub index_entry: &'a index::Entry,
+    /// The on-disk worktree path corresponding to this entry
+    pub worktree_path: PathBuf,
+    /// How this index entry changed
+    pub kind: ChangeKind,
+    /// file metadata that can be reused (optimization)
+    pub fstat: Option<std::fs::Metadata>,
+}
+
+#[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Debug)]
+///
+pub enum ChangeKind {
+    /// An index entry has no corresponding file in the worktree
+    Removed,
+    /// Ar new files that has been marked with git add but has not yet been
+    /// checked in yet. No diff is computed for these files because whatever is
+    /// on disk at commit time will be used
+    Added,
+    /// Called for files that may have changed in some form as indicated by `change`.
+    /// Note that this doesn't necessarily mean that the *content* of the file changed
+    /// see [`FileStatus`] for details
+    Modified {
+        /// How the file was modified exactly
+        modification: FileModification,
+        /// Whether this (changed) file also has an unresolved merge conflict
+        conflict: bool,
+    },
+    /// There are unresolved merge conflicts for this file
+    /// but it has not changed on disk
+    Conflict,
+}
+
+/// Computes the changes between the index and the worktree
+pub fn compare_to_index<'a: 'b, 'b>(
+    index: &'a index::State,
+    // TODO: use worktree cache instead
+    worktree: &'b Path,
+    capabilities: &'b fs::Capabilities,
+) -> impl Iterator<Item = Result<Change<'a>, Error>> + 'b {
+    // TODO: parallel with rayon
+    index.entries().iter().filter_map(|index_entry| {
+        let conflict = match index_entry.stage() {
+            0 => false,
+            1 => true,
+            _ => return None,
+        };
+        let git_path = index_entry.path(index);
+        if index_entry.flags.intersects(
+            index::entry::Flags::UPTODATE
+                | index::entry::Flags::SKIP_WORKTREE
+                | index::entry::Flags::ASSUME_VALID
+                | index::entry::Flags::FSMONITOR_VALID,
+        ) {
+            return None;
+        }
+
+        let path = if let Ok(path) = path::try_from_bstr(git_path) {
+            path
+        } else {
+            return Some(Err(Error::IllformedUtf8 {
```

Comment:
I think this `filter_map` should be restructured so that the core logic can return `Result<…>` so these acrobatics aren't needed.

Reply:
yeah I refacotred this pretty late, I used a visitor pattern previously. Probably better to move it to a function